src/Features/Core/Portable/Completion/CompletionItem.cs

```diff
@@ -84,6 +84,28 @@ public sealed class CompletionItem : IComparable<CompletionItem>
         /// </summary>
         internal Document Document { get; set; }
 
+        /// <summary>
+        /// The name of the <see cref="CompletionProvider"/> that created this 
+        /// <see cref="CompletionItem"/>. Not available to clients. Only used by 
+        /// the Completion subsystem itself for things like getting description text
+        /// and making additional change during commit.
+        /// </summary>
+        internal string ProviderName { get; set; }
+
+        /// <summary>
+        /// Whether to cache converted completion item for target editor. If returns <see langword="true"/>,
+        /// <see cref="CachedEditorCompletionItem"/> is used for storing the item.
+        /// Not available to clients. Only used to improve perf by Completion subsystem on a case-by-case basis.
+        /// </summary>
+        internal bool UseEditorCompletionItemCache { get; }
+
+        /// <summary>
+        /// Storing converted completion item for target editor. Used if <see cref="UseEditorCompletionItemCache"/> 
+        /// returns <see langword="true"/> 
+        /// Not available to clients. Only used to improve perf by Completion subsystem on a case-by-case basis.
+        /// </summary>
+        internal object CachedEditorCompletionItem { get; set; }
```

Comment:
if we feel liek this is hacky, then an alternative is to use a ConditionalWeakTable at the VS layer to map from CompletionItem to VSCompletionItem (without keeping CompletionItems alive longer than desired). But you'd have to test perf to make sure that didn't cause an problems.

Reply:
RE:Document issue, super llama comes to the rescue :D
https://github.com/dotnet/roslyn/pull/34952

Reply:
Once that's merged in and all sorted out then do grab a debugger and make sure we aren't leaking anything else with this cache. It'd be *very* easy to make a mistake.

Reply:
Sure. I will merge that change into this branch and make them a bundle deal.

Reply:
@jasonmalinowski @CyrusNajmabadi 
RE: ConditionalWeakTable, do you guys feel strongly about this? i.e. Do you want me to change it before merging, or it's something we can change later, after more profiling and better understanding of its potential perf impact for this feature?

Reply:
I'd feel moderately strongly about it at this point. With all the work that @CyrusNajmabadi and you have done making CompletionItem more immutable, it seems sad to immediately throw it out with this change.

I also still have some general reservations about this caching mechanism, because of the possibility of leaking editor data. The Editor's CompletionItem (for all we might know) might be capturing text buffers or views, where you might have to be caching these items on a per-view basis. Using a CWT means you can easily associate your cache with a view like that.

And yes, I don't have any _evidence_ that there's a concrete problem here with caching but I suspect we need to do careful testing to make sure we don't make any mistakes. It looks like we've already caught mistakes during this code review, so we clearly don't have the testing that we need for that.

Reply:
OK, I'm working on that right now